File: q0198/q0198.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1)
            return nums[0];
        else if (n == 2)
            return Math.max(nums[0], nums[1]);
        int[] maxMoney = new int[n];
        maxMoney[0] = nums[0];
        maxMoney[1] = Math.max(nums[0], nums[1]);
        for (int i = 2; i < n; i++) {
            int res1 = nums[i] + maxMoney[i - 2];
            int res2 = maxMoney[i - 1];
            maxMoney[i] = Math.max(res1, res2);
        }
        return maxMoney[n - 1];
    }

    public int robDimensionReduction(int[] nums) {
        int n = nums.length;
        if (n == 1)
            return nums[0];
        else if (n == 2)
            return Math.max(nums[0], nums[1]);

        int ans1 = nums[0];
        int ans2 = Math.max(nums[0], nums[1]);
        int res = 0;

        for (int i = 2; i < n; i++) {
            res = Math.max(ans1 + nums[i], ans2);
            ans1 = ans2;
            ans2 = res;
        }

        return res;
    }
}

File: q0198/q0198.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        elif n == 2:
            return max(nums[0], nums[1])
        maxMoney = [0] * n
        maxMoney[0] = nums[0]
        maxMoney[1] = max(nums[0], nums[1])
        for i in range(2, n):
            res1 = nums[i] + maxMoney[i - 2]
            res2 = maxMoney[i - 1]
            maxMoney[i] = max(res1, res2)
        return maxMoney[-1]

    # dimension reduction which achieves the same result
    def robDimensionReductio(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        elif n == 2:
            return max(nums[0], nums[1])

        ans1 = nums[0]
        ans2 = max(nums[0], nums[1])
        res = 0
        for i in range(2, n):
            res = max(ans1 + nums[i], ans2)
            ans1 = ans2
            ans2 = res
        return res

    ## using memoization
    def rob(self, nums):
        memo = {}

        def dp(i):
            if i == 0:
                return nums[0]
            if i == 1:
                return max(nums[0], nums[1])
            if i in memo:
                return memo[i]

            ans1 = nums[i] + dp(i - 2)
            ans2 = dp(i - 1)
            res = max(ans1, ans2)
            memo[i] = res
            return res

        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        return dp(len(nums) - 1)
